modules/table/src/main/java/org/apache/ignite/internal/table/distributed/StorageUpdateHandler.java

```diff
@@ -155,52 +190,82 @@ public void handleUpdateAll(
-                    BinaryRow row = entry.getValue() == null ? null : entry.getValue().binaryRow();
-
-                    locker.lock(rowId);
-
-                    performStorageCleanupIfNeeded(txId, rowId, entry.getValue() == null ? null : entry.getValue().commitTimestamp());
-
-                    if (commitTs != null) {
-                        storage.addWriteCommitted(rowId, row, commitTs);
-                    } else {
-                        BinaryRow oldRow = storage.addWrite(rowId, row, txId, commitTblId, commitPartId);
-
-                        if (oldRow != null) {
-                            assert commitTs == null : String.format("Expecting explicit txn: [txId=%s]", txId);
-                            // Previous uncommitted row should be removed from indexes.
-                            tryRemovePreviousWritesIndex(rowId, oldRow);
-                        }
-                    }
+        if (onApplication != null) {
+            onApplication.run();
+        }
+    }
 
-                    rowIds.add(rowId);
-                    indexUpdateHandler.addToIndexes(row, rowId);
+    private Entry<UUID, TimedBinaryRow> processEntriesUntilBatchLimit(
+            Entry<UUID, TimedBinaryRow> lastUnprocessedEntry,
+            UUID txId,
+            boolean trackWriteIntent,
+            @Nullable HybridTimestamp commitTs,
+            int commitTblId,
+            int commitPartId,
+            Iterator<Entry<UUID, TimedBinaryRow>> it,
+            int maxBatchLength,
+            boolean useTryLock
+    ) {
+        return storage.runConsistently(locker -> {
+            List<RowId> processedRowIds = new ArrayList<>();
+            int batchLength = 0;
+            Entry<UUID, TimedBinaryRow> entryToProcess = lastUnprocessedEntry;
+            while (entryToProcess != null) {
+                RowId rowId = new RowId(partitionId, entryToProcess.getKey());
+                BinaryRow row = entryToProcess.getValue() == null ? null : entryToProcess.getValue().binaryRow();
+                if (row != null) {
+                    batchLength += row.tupleSliceLength();
                 }
-
-                if (trackWriteIntent) {
-                    pendingRows.addPendingRowIds(txId, rowIds);
+                if (!processedRowIds.isEmpty() && batchLength > maxBatchLength) {
+                    break;
                 }
-
-                if (onApplication != null) {
-                    onApplication.run();
+                boolean rowProcessed = tryProcessRow(
+                        locker,
+                        commitTblId,
+                        commitPartId,
+                        rowId,
+                        txId,
+                        row,
+                        entryToProcess.getValue() == null ? null : entryToProcess.getValue().commitTimestamp(),
+                        commitTs,
+                        useTryLock
+                );
+                if (rowProcessed) {
```

Comment:
I don't see an `else` branch where you would return from this method. Are you sure that this code works if `tryLock` fails?